hints.hs

```diff
@@ -0,0 +1,309 @@
+isDeadSquare _ (DeadSquare _     True _    True _   )     = True -- space in corner
+isDeadSquare _ (DeadSquare _     True _    _    True)     = True -- "
+isDeadSquare _ (DeadSquare _     _    True True _   )     = True -- "
+isDeadSquare _ (DeadSquare _     _    True _    True)     = True -- "
+isDeadSquare b (DeadSquare (r,c) True _    _    _   )     = (nextWallLeft b (r,c) > (nextLifeLeft b (r,c))) &&
+                                                            (nextWallRight b (r,c) < (nextLifeRight b (r,c)))
+isDeadSquare b (DeadSquare (r,c) _    True _    _   )     = (nextWallLeft b (r,c) > (nextLifeLeft b (r,c))) && -- wall up/down cases equal
+                                                            (nextWallRight b (r,c) < (nextLifeRight b (r,c)))
+isDeadSquare b (DeadSquare (r,c) _    _    True _   )     = (nextWallUp b (r,c) > (nextLifeUp b (r,c))) &&
+                                                            (nextWallDown b (r,c) < (nextLifeDown b (r,c)))
+isDeadSquare b (DeadSquare (r,c) _    _    _    True)     = (nextWallUp b (r,c) > (nextLifeUp b (r,c))) && -- wall left/right cases equal
+                                                            (nextWallDown b (r,c) < (nextLifeDown b (r,c)))
+
+-- given board and coordinates,
+-- returns row or column of next wall in corresponding direction
+nextWallUp :: Board -> Coordinates -> Int
+nextWallUp b (r,c)
+    | char == '#' = r
+    | char == '.' = -1
+    | char == '*' = -1
+    | otherwise   = nextWallUp b (r-1,c)
+  where char = getCharacter b (r,c)
+
+nextWallDown :: Board -> Coordinates -> Int
+nextWallDown b (r,c)
+    | char == '#' = r
+    | char == '.' = 100000000000 -- something greater than the max level height
+    | char == '*' = 100000000000
+    | otherwise   = nextWallDown b (r+1,c)
+  where char = getCharacter b (r,c)
+
+nextWallLeft :: Board -> Coordinates -> Int
+nextWallLeft b (r,c)
+    | char == '#' = c
+    | char == '.' = -1
+    | char == '*' = -1
+    | otherwise   = nextWallLeft b (r,c-1)
+  where char = getCharacter b (r,c)
+
+nextWallRight :: Board -> Coordinates -> Int
+nextWallRight b (r,c)
+    | char == '#' = c
+    | char == '.' = 100000000000 -- something greater than the max level width
+    | char == '*' = 100000000000
+    | otherwise   = nextWallRight b (r,c+1)
+  where char = getCharacter b (r,c)
+
+-- given board and cooredinates,
+-- returns row or column of next 'life' space in the corresponding direction
+-- ('life' means there is no deadlock -- the wall ends (' ' or '.') on both sides (so player can push box away))
+nextLifeUp :: Board -> Coordinates -> Int
+nextLifeUp b (r,c)
+    | player == '#'                      = r-1 -- if one exists, it is above the wall (unreachable)
+    | ((left == ' ') || (left == '.')) &&
+      ((right == ' ') || (right == '.')) = r
+    | otherwise                          = nextLifeUp b (r-1,c)
+  where player = getCharacter b (r,c)
```

Comment:
By player, did you mean a box here? My bad if I'm misunderstanding something

Reply:
Yeah, my bad. Good catch